### PR TITLE
Implementation of new config subcommands

### PR DESCRIPTION
**Pull request by denniss**
Implement:
https://github.com/bundler/bundler/issues/4600

### CODE REVIEW

File: spec/commands/config2_spec.rb [hunk 1]
```diff
@@ -83,13 +83,155 @@
...
+      File.open(bundled_app(".bundle/config"), "w") do |f|
+        f.write 'BUNDLE_FOO: "$BUILD_DIR"'
+      end
+
+      bundle "config2 set bar baz"
+      run "puts Bundler.settings.send(:local_config_file).read"
+
+      # Starting in Ruby 2.1, YAML automatically adds double quotes
+      # around some values, including $ and newlines.
+      expect(out).to include('BUNDLE_FOO: "$BUILD_DIR"')
```
Comment (segiddins) [inline spec/commands/config2_spec.rb]: we use our own YAML serializer, so this shouldn't be version-dependent?

**Comment by colby-swandale** [comment]
@denniss can you fix the rubocop feedback reported in travis please 😄

**Comment by denniss** [comment]
@indirect Can you take a peek at this? I started this in response to https://github.com/bundler/bundler/issues/4600 . I used the existing `config_spec.rb` as a starting point and to pass the implementation of the new command (currently called `config2` and the spec is in `config2_spec.rb`).

**Comment by colby-swandale** [comment]
ping @indirect

File: spec/commands/config2_spec.rb [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+      bundle "config2 --global set foo global"
+      expect(out).to match(/Your application has set foo to "local"/)
+
+      run "puts Bundler.settings[:foo]"
+      expect(out).to eq("local")
+    end
+
+    it "can be unset" do
+      bundle "config2 unset --global foo global"
```
Comment (indirect) [inline spec/commands/config2_spec.rb]: looks like this was supposed to be a set?

**Comment by denniss** [comment]
Changed the specs to not pass `--local` by default. Also fixed the mistake you found @indirect

File: spec/commands/config2_spec.rb (hunk 1 above)
Comment (segiddins) [inline spec/commands/config2_spec.rb]: ping on this comment?

Comment (denniss) [inline spec/commands/config2_spec.rb]: Removed the ruby version specifier

**Comment by denniss** [comment]
Currently new implementation is aliased as `config2`. I'd like to remove the existing `config` command and renamed `config2` to `config` if you guys think this looks okay.

**Comment by indirect** [comment]
Since this is a breaking change, we can merge it to master behind a feature flag that switches the new implementation on. Then, after we ship 2.0 and the feature flag is always on, we can delete the old implementation. 👍

**Comment by denniss** [comment]
@indirect @segiddins While working through merge conflight, I was wondering: If the same key exist for environment variable (`ENV['BUNDLE_BAR']) and global (~/.bundle/config) which has higher precedence? Currently environment variable has higher precedence. Is this correct? Does it matter?

**Comment by segiddins** [comment]
> Currently environment variable has higher precedence. Is this correct? Does it matter?
yes, and yes

**Comment by segiddins** [comment]
@denniss what's the status of this PR?

**Comment by denniss** [comment]
I won't be able to touch this for another 2 weeks. Busy with life at the
moment. Mich apologies.

**Comment by segiddins** [comment]
https://github.com/bundler/bundler/pull/5981 is an alternative to this PR

**Comment by segiddins** [comment]
Superseded by #5981

**Pull request closed without merging**